contracts/test/recovery-test.ts

```diff
@@ -133,6 +137,55 @@ describe("Recovery", async function () {
     expect(await blsWallet.recoveryHash()).to.equal(newRecoveryHash);
   });
 
+  it("should set recovery hash using client bls wallet wrapper function", async function () {
+    // set instantly from 0 value
+    const trustedWalletAddress = "0x7321d1D33E94f294c144aA332f75411372741d33";
+    const walletHash = await vg.hashFromWallet(wallet3.address);
+    const salt = "test salt";
+    const saltBytes32String = ethers.utils.formatBytes32String(salt);
+    const recoveryHash = ethers.utils.solidityKeccak256(
+      ["address", "bytes32", "bytes32"],
+      [trustedWalletAddress, walletHash, saltBytes32String],
+    );
+
+    const bundle = await wallet3.getBundleSetRecoveryHash(
+      "test salt",
+      trustedWalletAddress,
+    );
+    const bundleTxn = await fx.verificationGateway.processBundle(bundle);
+    await bundleTxn.wait();
+
+    expect(await blsWallet3.recoveryHash()).to.equal(recoveryHash);
+  });
+
+  it("should recover blswallet via blswallet to new bls key using bls client module", async function () {
+    // Set recovery hash
+    const wallet4 = await fx.lazyBlsWallets[3]();
+    const bundle = await wallet4.getBundleSetRecoveryHash(
+      "test salt",
+      wallet3.address,
+    );
+    const bundleTxn = await fx.verificationGateway.processBundle(bundle);
+    await bundleTxn.wait();
+
+    // Recover wallet
+    const newPrivateKey = wallet3.getRandomBlsPrivateKey();
+    const recoveryBundle = await wallet3.getBundleRecoverWallet(
+      wallet4.address,
+      newPrivateKey,
+      "test salt",
+      fx.verificationGateway,
+    );
+    const recoveryBundleTxn = await fx.verificationGateway.processBundle(
+      recoveryBundle,
+    );
+    await recoveryBundleTxn.wait();
+
+    const newHash = wallet4.blsWalletSigner.getPublicKeyHash(newPrivateKey);
+    expect(await vg.hashFromWallet(wallet4.address)).to.eql(newHash);
+    expect(await vg.walletFromHash(newHash)).to.eql(wallet4.address);
```

Comment:
General recovery question - my understanding of this scenario is that wallet 3 should not be able to send transactions after being recovered to wallet 4, is that correct?

If so, do you see value in a test that checks wallet 3 cannot send transactions after being recovered? That might be covered by existing assertions/not be desirable, curious on your thoughts here.

Reply:
So in this situation, wallet 3 is actually the trusted wallet for wallet 4.  I've been using the term trusted wallet recently because I think that makes more sense then what I've been calling it.  In this situation, we can assume that it was actually wallet 4 that lost it's private key and needs to recover to a new one.  Wallet 3 is the trusted wallet that can call the recover wallet function to swap the PK for wallet 4.  

After the recovery is done, wallet 4's old private key should no longer be able to send txs.  But the new private key that now maps to wallet 4's public key should be able to send transactions for that hash.  That's why here I check that the public key hash for the new private key is the same as the hash that maps to wallet 4's address.  If that's the case, the new private key can send txs for that address.

With regard to the test.  We already have a decent amount of recovery tests in here which is why I just wanted to add a test to check that the bundle created from the new functions works.  I just looked though and it doesn't look like we have a test that explicitly checks that the old private key doesn't work anymore.  So I think that would be a valuable test.

I could add an extra line to this test to check that the old private key no longer maps to any public key hash.  Which would mean it can't send any txs.  When a recovery happens we just zero out the hash that the private key maps to in the mapping stored in the verification gateway. So that private key can't do anything anymore.

Reply:
Yeah think I was confused about how this was working. I've got a good grasp now, thanks for clearing that up!

Trusted wallet as a term also makes sense. Wondering if renaming `blsWallet3` to something like `trustedBlsWallet` would improve readability?

Makes sense in regards to additional test case.